python/lsst/ip/isr/isrMock.py

```diff
@@ -0,0 +1,915 @@
+        ----------
+        amp : `lsst.afw.image.ImageF`
+            Amplifier image to use for conversions.
+        x : int
+            X-coordinate of the point to transform.
+        y : int
+            Y-coordinate of the point to transform.
+
+        Returns
+        -------
+        u : int
+            Transformed x-coordinate.
+        v : int
+            Transformed y-coordinate.
+
+        Notes
+        -----
+        The output is transposed intentionally here, to match the
+        internal transpose between numpy and afw.image coordinates.
+        """
+        if self.config.isTrimmed:
+            u = x + amp.getBBox().getBeginX()
+            v = y + amp.getBBox().getBeginY()
+        else:
+            u = x + amp.getRawDataBBox().getBeginX()
+            v = y + amp.getRawDataBBox().getBeginY()
+
+        return (v, u)
+
+    # Simple data values.
+    def amplifierAddNoise(self, ampData, mean, sigma):
+        r"""Add Gaussian noise to an amplifier's image data.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        mean : float
+            Mean value of the Gaussian noise.
+        sigma : float
+            Sigma of the Gaussian noise.
+        """
+        ampArr = ampData.getArray()
+        ampArr[:] = ampArr[:] + self.rng.normal(mean, sigma,
+                                                size=ampData.getDimensions()).transpose()
+
+    def amplifierAddYGradient(self, ampData, start, end):
+        r"""Add a y-axis linear gradient to an amplifier's image data.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        start : float
+            Start value of the gradient (at y=0).
+        end : float
+            End value of the gradient (at y=ymax).
+        """
+        nPixY = ampData.getDimensions().getY()
+        ampArr = ampData.getArray()
+        ampArr[:] = ampArr[:] + (np.interp(range(nPixY), (0, nPixY - 1), (start, end)).reshape(nPixY, 1) +
+                                 np.zeros(ampData.getDimensions()).transpose())
+
+    def amplifierAddSource(self, ampData, scale):
+        r"""Add a single Gaussian source to an amplifier.
+
+         This method operates in the amplifier coordinate frame.
+
+        Parameters
+        ----------
+        ampData : `lsst.afw.image.ImageF`
+            Amplifier image to operate on.
+        scale : float
+            Peak flux of the source to add.
+        """
+        for x in range(0, ampData.getDimensions().getX()):
+            for y in range(0, ampData.getDimensions().getY()):
+                ampData.getArray()[y][x] = (ampData.getArray()[y][x] +
+                                            scale * np.exp(-0.5 * ((x - 50)**2 + (y - 25)**2)**2 / 3.0**2))
```

Comment:
I've not noticed speed issues, but this could easily loop only over the region the source is significant.  The placement was arbitrary, but putting a real signal near the defect does (optionally) test that defect interpolation looks reasonable.  With the sharp star (a think-o more than a typo), the interpolated region only covers the sky noise, but I'll check when running the tests.

Reply:
Yeah, operating on 1Mpix instead of 16Mpix isn't too bad, especially if you're not doing it multiple times. I think the star sharpness was the main point here, so as long as that's fixed and it's not problematically slow it should be fine (though I don't really know where this stuff will be used, and that certainly drives how much we should care about speed).

Reply:
My intention was that this would be used for unit tests as well as for development when real data is either unavailable or has confounding signals.  I'm not expecting this to ever be called in regular processing, so I haven't been worried about speed.